### DISCUSSION TITLE
How to debug a never-returning open()?

### DISCUSSION THREAD

**Discussion by matthiasbeyer** [tokio-rs/tokio#3240] [Q&A]
Hi,

I have an call to `open()` that does not return.

My code:

```rust
        trace!("Reading to buffer: {}", p.display());
        let mut buf = vec![];
        let mut file = tokio::fs::OpenOptions::new()
            .create(false)
            .create_new(false)
            .write(false)
            .read(true)
            .open(&p)
            .await?;

        trace!("Opening worked...");
        file.read_to_end(&mut buf)
            .await
            .with_context(|| anyhow!("Reading file: {}", p.display()))?;
        trace!("Reading worked...");
```

I get the log output from "Reading to buffer" with the path, the path exists on my FS.

strace reports this:

```
statx(AT_FDCWD, "/tmp/REDACTED", AT_STATX_SYNC_AS_STAT, STATX_ALL, {stx_mask=STATX_BASIC_STATS, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=5, ...}) = 0
write(2, "\33[0m\33[38;5;8m[\33[0m2020-12-09T13:"..., 189) = 189
write(2, "\33[0m\33[38;5;8m[\33[0m2020-12-09T13:"..., 190) = 190
futex(0x56166ce20df8, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x56166ce20da0, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x56166ce1fa88, FUTEX_WAIT_PRIVATE, 4294967295, NULL)
```

and blocks there (I waited up to 1:30 min). The file is less than 10 bytes.

How to debug this?

**Comment by matthiasbeyer** [comment]
(btw I also tried `tokio::fs::read(&p)`, but that has the exact same issue... so I tried the above code instead...

**Comment by Darksonn** [comment]
It sounds like your runtime is configured with [`max_threads`][1] equal to [`core_threads`][2]. When this happens, the runtime is not able to execute `spawn_blocking` calls since there are no threads left for blocking operations.

[1]: https://docs.rs/tokio/0.2/tokio/runtime/struct.Builder.html#method.max_threads
[2]: https://docs.rs/tokio/0.2/tokio/runtime/struct.Builder.html#method.core_threads

**Reply by matthiasbeyer** [reply]
That doesn't add up to me.
Because wouldn't that mean that other stuff would still be running? If that was the case, the `strace` output would still yield lines, wouldn't it?
Also, I have a i7 with 4+4 cores... and the quoted lines are the only future that runs... so I doubt that there is more than one (blocking) future at that time...

**Reply by Darksonn** [reply]
Well async code can still run. How did you set up your runtime?

**Reply by matthiasbeyer** [reply]
```
#[tokio::main]
async fn main() -> Result<()>
```

Is that what you mean?

**Reply by Darksonn** [reply]
I guess that's not the reason then. When I run your code locally, it works fine.
